File: CtCi/ch01/Exercise1_1.cpp
#include <cstdio>


bool is_unique(char *str) {
    // assume that string chars fit in ascii table
    int idx = 0, c;
    long long first_half = 0, second_half = 0, checker;
    while (c = str[idx++]) {
        if (c < 64) {
            checker = first_half;
            first_half |= (1 << c);
        } else {
            c %= 64;
            checker = second_half;
            second_half |= (1 << c);
        }
        if (checker & (1 << c)) return false;
    }
    return true;
}


int main() {
    char str[30];
    scanf("%s", str);
    printf(is_unique(str) ? "true\n" : "false\n");

    return 0;
}

File: CtCi/ch01/Exercise1_1.java
import java.lang.*;
import java.io.*;

class Exercise1_1 {

    public static Boolean isUnique(String word) {
        long firstHalf = 0, secondHalf = 0, checker;
        for (int i = 0; i < word.length(); i++) {
            int c = (int)word.charAt(i);
            if (c < 64) {
                checker = firstHalf;
                firstHalf |= (1 << c);
            } else {
                c %= 64;
                checker = secondHalf;
                secondHalf |= (1 << c);
            }
            if ((checker & (1 << c)) > 0) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(Exercise1_1.isUnique("gustavo lopes"));
    } 
}